staging/src/k8s.io/client-go/openapi/client.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/kube-openapi/pkg/handler3"
+)
+
+type Client interface {
+	Paths() (map[string]GroupVersion, error)
+}
+
+type client struct {
+	// URL includes the `hash` query param to take advantage of cache busting
+	restClient     rest.Interface
+	discoveryToken sync.Once
+	discoveryMap   map[string]GroupVersion
+	discoveryError error
+}
+
+func NewClient(restClient rest.Interface) Client {
+	return &client{
+		restClient: restClient,
+	}
+}
+
+func (c *client) Paths() (map[string]GroupVersion, error) {
+	c.discoveryToken.Do(func() {
+		data, err := c.restClient.Get().
```

Comment:
You could have possibly split this into a "CachedClient" and a "NonCachedClient" (with NonCached doing what inside the do method, and the Cached doing the sync.Once). I can see use-cases where we would want to have a non-cached version.